### PR TITLE
add pyobjcryst

### PR DESCRIPTION
**Pull request by CJ-Wright**
This requires scons to build, does this mean that it will only be built for py27?

### CODE REVIEW

File: recipes/pyobjcryst/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+build:
+  number: 0
+  skip: True  # [win]
+
+requirements:
+  build:
+    - python
+    - setuptools
+    - numpy 1.11.2
```
Comment (CJ-Wright) [inline recipes/pyobjcryst/meta.yaml]: @pavoljuhas Do we need to pin to this numpy or can we go higher?

File: recipes/pyobjcryst/build.sh [hunk 2]
```diff
@@ -0,0 +1,21 @@
+#!/bin/bash
+
+export CPATH="${PREFIX}/include:$CPATH"
+export LIBRARY_PATH="${PREFIX}/lib:$LIBRARY_PATH"
+
+MYNCPU=$(( (CPU_COUNT > 4) ? 4 : CPU_COUNT ))
+
+if [ `uname` == Darwin ]; then
+    export DYLD_FALLBACK_LIBRARY_PATH="${PREFIX}/lib"
+fi
```
Comment (isuruf) [inline recipes/pyobjcryst/build.sh]: Add `toolchain` as a build requirement and you can remove the lines above (except for `MYNCPU`)

**Comment by pavoljuhas** [comment]
> This requires scons to build, does this mean that it will only be built for py27?

It can also built for 3.4, 3.5, 3.6, but scons executable needs to be from another non-build environment.  Scons scripts in pyobjcryst build for the first `python` in the PATH or for a python instance given on command line as `scons python=/for/this/python`.  The Python process running scons can be different from Python for which the package is built.  The travis script for the pyobjcryst package sets it up [here](https://github.com/diffpy/pyobjcryst/blob/0f20976d4a00c215bf81292a4798f709e7e63012/.travis.yml#L64-L66).

If a global scons program is too tricky to configure on conda-forge, just install with
```
python setup.py install
```
This works the same, except it has no `-jN` switch to build C++ files in parallel.

File: recipes/pyobjcryst/meta.yaml (hunk 1 above)
Comment (pavoljuhas) [reply recipes/pyobjcryst/meta.yaml]: Runtime numpy can be more recent; it will work with pyobjcryst build against numpy 1.11.2.  On the other hand, pyobjcryst build against the last numpy may be incompatible with earlier numpy-s.  I don't want pyobjcryst forcing the latest numpy on its target environment.

Please, keep the build-stage numpy as is, but use `numpy >=1.11.2` for the `run` requirements.

**Comment by CJ-Wright** [comment]
@isuruf any thoughts on the `scons` 2 vs. 3 issue?

File: recipes/pyobjcryst/SConstruct.patch [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
++++ b/SConstruct
+@@ -97,11 +97,6 @@ good_python_flags = lambda n : (
+ pyversion = pyoutput('import sys; print("%i.%i" % sys.version_info[:2])')
+ pythonconfig = 'python%s-config' % (pyversion if pyversion[0] == '3' else '')
+ # Verify python-config comes from the same path as the target python.
+-xpython = env.WhereIs(env['python'])
+-xpythonconfig = env.WhereIs(pythonconfig)
+-if os.path.dirname(xpython) != os.path.dirname(xpythonconfig):
+-    print("Inconsistent paths of %r and %r" % (xpython, xpythonconfig))
+-    Exit(1)
```
Comment (pavoljuhas) [inline recipes/pyobjcryst/SConstruct.patch]: @CJ-Wright - Is that change necessary?  That check prevents usage of incorrect python-config which could be tricky to track down when things go haywire.

Can you point me to a build log that failed because of the check?

Comment (CJ-Wright) [reply recipes/pyobjcryst/SConstruct.patch]: this one? https://circleci.com/gh/conda-forge/staged-recipes/22905?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

Comment (pavoljuhas) [reply recipes/pyobjcryst/SConstruct.patch]: I see.  Can you replace that patch with [0001-BLD-in-scons-convert-python-output-to-strings.patch.txt](https://github.com/conda-forge/staged-recipes/files/1679301/0001-BLD-in-scons-convert-python-output-to-strings.patch.txt) ?
(also at https://github.com/pavoljuhas/pyobjcryst/commit/35ec0dbd47c24ecb43c1fb8b16d256cb3f805ccf)

The check failed, because `"3" == b"3"` evaluated to False when checking Python version.

**Comment by CJ-Wright** [comment]
@pavoljuhas ran into another error it seems to think that `libobjcryst` although it is conda installed

**Comment by pavoljuhas** [comment]
@CJ-Wright - can you try with the current master from diffpy/pyobjcryst?

If that fails, using `python setup.py install` is another option.

**Comment by pavoljuhas** [comment]
[failed circleci](https://goo.gl/nWZ7AV) -- looks like this needs `CPATH` and `LIBRARY_PATH` exports that were in build.sh.

**Comment by CJ-Wright** [comment]
This doesn't build on windows so @conda-forge/staged-recipes ready for review!

File: recipes/pyobjcryst/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+package:
+  name: {{ name|lower }}
+  version: {{ version }}
+
+source:
+  fn: {{ name }}-{{ version }}.tar.gz
+  url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.tar.gz
+  sha256: {{ sha256 }}
+  patches:
+    - SConstruct.patch
```
Comment (pavoljuhas) [inline recipes/pyobjcryst/meta.yaml]: SConstruct.patch is now not necessary, because scons is not used.  You can also remove `scons` from build requirements.

File: recipes/pyobjcryst/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+build:
+  number: 0
+  skip: True  # [win or py2k or osx]
```
Comment (pavoljuhas) [inline recipes/pyobjcryst/meta.yaml]: This should build on `osx`, why skip there?

File: recipes/pyobjcryst/meta.yaml (hunk 1 above)
Comment (dougalsutherland) [reply recipes/pyobjcryst/meta.yaml]: Not a big deal, but what about `numpy 1.11.*,>=1.11.2` assuming the `.2` patch release is necessary to build?

File: recipes/pyobjcryst/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+  script: python setup.py install
+
+requirements:
+  build:
+    - python
+    - setuptools
+    - numpy 1.11.2
+    - libobjcryst 2015.1.*  # [py2k]
```
Comment (dougalsutherland) [inline recipes/pyobjcryst/meta.yaml]: We're skipping `[py2k]` anyway, no?

File: recipes/pyobjcryst/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+test:
+  imports:
+    - pyobjcryst
+    - pyobjcryst.tests
+
+about:
+  home: https://github.com/diffpy/pyobjcryst
+  license: BNL
+  license_family: BSD
```
Comment (dougalsutherland) [inline recipes/pyobjcryst/meta.yaml]: Is this actually considered a BSD-family license? Looks kind of different on a skim.

Comment (CJ-Wright) [reply recipes/pyobjcryst/meta.yaml]: Attn: @pavoljuhas

File: recipes/pyobjcryst/meta.yaml (hunk 5 above)
Comment (CJ-Wright) [reply recipes/pyobjcryst/meta.yaml]: OSX build is failing on travis I thought we could debug it in the feedstock though

**Comment by pavoljuhas** [comment]
The osx package is for some reason built twice on travis.

The [first build](https://travis-ci.org/conda-forge/staged-recipes/builds/335631803#L221) links with conda-forge boost and [passes](https://travis-ci.org/conda-forge/staged-recipes/builds/335631803#L559).

The [second build](https://travis-ci.org/conda-forge/staged-recipes/builds/335631803#L580) links with boost from the defaults channel and [fails](https://travis-ci.org/conda-forge/staged-recipes/builds/335631803#L921).  Any idea why is that second build there?

File: recipes/pyobjcryst/meta.yaml (hunk 7 above)
Comment (pavoljuhas) [reply recipes/pyobjcryst/meta.yaml]: The license text is from Intellectual Property office at the BNL and contains licenses of the previous projects we build on.  The 3-clause BSD is sprinkled in the text at

https://github.com/diffpy/pyobjcryst/blob/35ec0dbd47c24ecb43c1fb8b16d256cb3f805ccf/LICENSE.txt#L51-L68
and
https://github.com/diffpy/pyobjcryst/blob/35ec0dbd47c24ecb43c1fb8b16d256cb3f805ccf/LICENSE.txt#L79-L98

**Comment by pavoljuhas** [comment]
@CJ-Wright - let's test with package unit tests.  Please merge https://github.com/conda-forge/staged-recipes/commit/1e633307d61c5a60cca436fb2d37347c36407f51.

**Comment by CJ-Wright** [comment]
@pavoljuhas Can you push (or PR) to my branch?

**Comment by CJ-Wright** [comment]
The failure on travis was/is due to boost 1.61 not being packaged for py36 on conda-forge.
Can we bump the boost version?
If not I guess we need to put in a PR to boost to build 1.61 for py36
or
not build on py36 (I'd prefer not to do this)

**Comment by pavoljuhas** [comment]
> Can we bump the boost version?

Sure that should work fine on conda-forge.

The older boost is for building with stock Anaconda, because their most recent packages link with their own compiler toolchain and C/C++ libraries.  I have not yet figured how to build w/r to that.  This is however no issue for conda-forge.  The current cf boost is good.

For the sake of consistency, please use the same boost version also for libobjcryst.

**Comment by dougalsutherland** [comment]
`conda-forge` recipes in general should be using `boost 1.65.1` (as noted [in the pinnings](https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/master/recipe/conda_build_config.yaml#L7-L8)).

**Comment by CJ-Wright** [comment]
Pending https://github.com/conda-forge/libobjcryst-feedstock/pull/1

**Comment by pavoljuhas** [comment]
@dougalsutherland - is there a way to have our recipe include those pinning rules?

**Comment by CJ-Wright** [comment]
@pavoljuhas it seems that new libobjcryst/boost broke things :confused:

**Comment by dougalsutherland** [comment]
@pavoljuhas Sometime in the not-too-distant future, it’ll be automatic. For now, you have to do them manually for the ones you link against.

**Comment by pavoljuhas** [comment]
@CJ-Wright - that [build error](https://travis-ci.org/conda-forge/staged-recipes/builds/335903782#L373) should be fixed by diffpy/pyobjcryst@0f20976d4a00c215bf81292a4798f709e7e63012.
Can you add it to the recipe as a patch?
[I don't have time to release a bumped-up pyobjcryst this week.]

**Comment by CJ-Wright** [comment]
This doesn't build on windows so @conda-forge/staged-recipes ready for revew/merge.
@pavoljuhas are you ok with this PR?
